sys/net/sock: add sock_udp_sendv() API

benpicco opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md.
-->

### Contribution description

Both GNRC and LWIP support the notion of packet snips/slices: A packet does not have to be in continuous memory.
This is useful when e.g. sending a blockwise CoAP message where the CoAP header can be a small buffer and the CoAP block payload can be just a pointer into a larger buffer (possibly on ROM).

However, the sock API completely throws this out of the window by only allowing a single payload buffer in `sock_udp_send()`.
This means we need another buffer to copy the payload and the header into, artificially doubling memory requirements and introducing an unnecessary copy step. 

Fix this by introducing a new `sock_udp_sendv()` function. This allows to specify an array of payload chunks instead of a single buffer. `sock_udp_send()` is now simple a wrapper around the new function.

#### Implemented for

 - [x] GNRC
 - [x] LWIP
 - [x] OpenWSN

### Testing procedure

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

maribu commented on sys/include/net/sock/udp.h:
```diff
@@ -352,6 +352,14 @@ typedef struct {
+typedef struct {
+    const void *data;   /**< payload      */
+    size_t len;         /**< payload size */
+} sock_udp_data_snip_t;
```
Why not use https://api.riot-os.org/iolist_8h.html ?

This is also what is presented at very bottom of the stack to the netdev, so using it here could make some nice optimisations possible. Also, since that type is bound to be used by drivers anyway, reusing the helper functions if that type could safe some rom of the same functionality is needed by both driver and application.

miri64 replied:
This would make `sock` unusable outside the RIOT-context, though, which used to be a design goal (not sure if with the most recent changes this is still the case though). One should be able to use `sock` on Linux as well (this was how nanocoap was tested back then).

benpicco replied:
Hm `iolist_t` implies a linked list which makes this a bit more cumbersome to use.
It is also translated into `gnrc_pktsnip_t` down the stack, so we don't benefit from type reuse.

benpicco replied:
> This would make `sock` unusable outside the RIOT-context

Do you mean with `iolist` or with the new API? Last time I checked there is no `sock_udp_send_aux()` on Linux, so we'd need a wrapper either way.

maribu replied:
Does iol contain anything that would be difficult to provide outside of RIOT?

Also, isn't that use case covered already by the `native` board?

maribu replied:
> Hm `iolist_t` implies a linked list which makes this a bit more cumbersome to use. It is also translated into `gnrc_pktsnip_t` down the stack, so we don't benefit from type reuse.

Couldn't you just enque them and adapt the free function to stop the iteration after the transport layer snip? The pktsnips have iolist as "base class" structure anyway.

benpicco replied:
> The pktsnips have iolist as "base class" structure anyway.

No [they don't](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/gnrc/pkt.h#L108).

maribu replied:
It would turn enquing them into O(1), but given that it is unlikely that there are more than two application layer fragments, it seems not to be worth the effort.

maribu reviewed (commented):
I do like the idea very much. One comment inline.

kaspar030 reviewed (changes requested):
Why an array? There's reasons why gnrc and netdev use iolists instead of arrays. They're much more stack friendly, especially for layered protocols.
And, iolists are used at the lowest level. I don't think gnrc can do it, because it needs pktsnips, but theoretically an application could provide an iolist, and the network stack just has to prepend the udp header and pass it on. Maybe this can be added to gnrc's machinery, with a special "iolist" pktsnip type.
An array would need to be reallocated, potentially multiple times.

benpicco commented:
We are on the sock layer here so we can't use network stack native list representations anyway.
I don't see how lists or arrays differ in stack usage (in fact the list will store an additional pointer, so 4 more bytes per element) but lists are more awkward to use:

 - Both GNRC and LWIP expect the last element to be written first. This would either mean that the user has to supply the list in an 'unnatural' order, use double-linked lists or we reverse the list inside the network stack
 - Array can be statically initialized (and be compile time `const`) 

I agree that lists provide additional flexibility in case someone wants to add multiple layers on top of `sock` though.

kaspar030 commented:
> I don't see how lists or arrays differ in stack usage (in fact the list will store an additional pointer, so 4 more bytes per element)

Think of an application writing some content using protocol X. This is a potential call stack using iolists:

```
proto_X_send(..., buf, len) {
  iolist_t entry = { .iol_data=buf, iol_len=len }

 proto_X_udp_send(..., &entry);
}

proto_X_udp_send(..., iolist_t entry) {
 buf, len = proto_X_create_hdr(...);
 iolist_t proto_x = {.iol_data = buf, .iol_len=len, iol_next=entry };
udp_sendv(..., &proto_x);
}

udp_sendv(..., &proto_x) {
buf, len = udp_build_hdr(...);
 iolist_t udp = {.iol_data = buf, .iol_len=len, iol_next=proto_x };
ip_send(..., &udp);
}

...
```

Using arrays, each layer would probably get an array as parameter. to add data, that array either needs to be in reverse order and large enough (who knows how large?), or, be recreated *on each layer*, but again, how big? variable length?.

IMO, lists save a lot here.

benpicco commented:
How about this? I'm not entirely happy with having to reverse the list depending on which network stack is used, but then again this is neatly hidden from the user and the list based interface does indeed provide more flexibility to the application.

kaspar030 commented:
I still don't understand why the iolist has to be reversed. just to be clear, iolists are in-order, meaning, first entry (list head) contains bytes 0-x, second contains x+1 . ..., and so on, right?

That's how it is passed from the application, and that's how the good old netdev expects (or expected) it.

In the PR, you're reversing that for lwip and openwsn, but not gnrc? Even though lwip writes in-order using netcon_write_partly() and openwsn also copies in-order? does gnrc_pktbuf_add prepend?

It seems to me something is turned around. if gnrc_pktbuf_add prepends, we need something that appends. There's gnrc_pkt_append, that sounds like it could be used (but might need allocation of a pktsnip, even though the data could be just pointed to?

but, are we clear on the supposed order of the iolists?

(You really don't want to re-use iolist.h?)

benpicco commented:
Sometimes one does not see the forest due to all the trees :smile: 
I've added `gnrc_pktbuf_add_tail()` and reworked the buffer chaining in `lwip_sock_sendv()` so now we can always use the buffer list in the right order.

I'll split up the commits and force push shortly.

kaspar030 commented on pkg/lwip/contrib/sock/lwip_sock.c:
```diff
@@ -616,22 +598,20 @@ ssize_t lwip_sock_sendv(struct netconn *conn, sock_tx_snip_t *snips,
-        buf = netbuf_new();
-        if ((buf == NULL) || (netbuf_alloc(buf, snip->len) == NULL)) {
+    for (const sock_tx_snip_t *snip = snips; snip != NULL; snip = snip->next) {
+        struct netbuf *tail = netbuf_new();
```
seeing this. is the data copied? if that's the case, we could allocate the whole pkt, then write the snips in there. saves allocations and thus overhead.

benpicco replied:
tbh this surprised me, doesn't LWIP support the concept of memory chunks? It certainly does for RX - maybe @miri64 or @yarrick know how to do this without copying the data

benpicco replied:
see 1bad89cf6194a1a34ef66181df6b57a0792e34fa

miri64 commented on sys/include/net/gnrc/pktbuf.h:
```diff
@@ -94,6 +94,44 @@ void gnrc_pktbuf_init(void);
+ * @return  0 on success
+ * @return  -ENOMEM, if no space is left in the packet buffer.
+ */
+static inline int gnrc_pktbuf_add_tail(gnrc_pktsnip_t **head,
+                                       const void *data, size_t size,
+                                       gnrc_nettype_t type)
+{
+    gnrc_pktsnip_t *snip = gnrc_pktbuf_add(NULL, data, size, type);
+
+    if (snip == NULL) {
+        return -ENOMEM;
+    }
+
+    *head = gnrc_pkt_append(*head, snip);
+
+    return 0;
```
Ok [after having some stomach aches after a conversation in the Matrix chat], I am fine with this change. It is just a convenience function (which might also be usable elsewhere). Does it make sense to emphasize that the actual list order is meant, not the packet order (which is reversed in the reception case).

benpicco replied:
Actually with 7a5a5a3b7eecc18b5e0b4586a4850e1acab18b56 we don't need this function anymore. (It could still be handy though).

The question is if we want to modify the `gnrc_pktsnip_t` like that - ideally we could have a blocking future where we don't have to copy user data into GNRC in sdsdssdsdsternal buffers, but just keep the pointer to it while sending.

However until then copying the multiple user snips into one GNRC snip saves some allocation overhead.

yarrick commented on pkg/lwip/contrib/sock/lwip_sock.c:
```diff
@@ -625,13 +634,19 @@ ssize_t lwip_sock_send(struct netconn *conn, const void *data, size_t len,
     else if (tmp->type & NETCONN_TCP) {
-        err = netconn_write_partly(tmp, data, len, 0, (size_t *)(&res));
+        while (snips) {
+            err = netconn_write_partly(tmp, snips->data, snips->len, 0, (size_t *)(&res));
```
Can we not use the buf created above here? And just do one send?

benpicco replied:
Yea I was wondering why `buf` was not used at all in the TCP case.
I suppose we can do 

```C
netconn_write_partly(tmp, buf->p->payload, buf->p->len, 0, (size_t *)(&res));
```

here

yarrick reviewed (commented):
I am not familiar with the netconn/netbuf API, but it seems the TCP send can be simplified.

yarrick commented on pkg/lwip/contrib/sock/lwip_sock.c:
```diff
@@ -640,13 +640,7 @@ ssize_t lwip_sock_sendv(struct netconn *conn, const sock_tx_snip_t *snips,
-            }
-            snips = snips->next;
-        }
+        err = netconn_write_partly(tmp, buf->p->payload, buf->p->len, 0, (size_t *)(&res));
```
The pbuf inside might be linked as well, so this will only send the contents of the first buffer. Hmm

benpicco replied:
It was however just allocated [a few lines above](https://github.com/RIOT-OS/RIOT/blob/5cb7ee8338d9e678107ecebc4d13798b4dc5493e/pkg/lwip/contrib/sock/lwip_sock.c#L601)

yarrick replied:
There is a `netconn_write_vectors_partly` which takes an array of `struct netvector` (containing ptr/len). That is probably the best (as loop of write_partly might send some data out of order)

yarrick replied:
netbuf internal pbuf is allocated with `PBUF_RAM` so it is one contiguous buffer. So the code as it is now should be fine.

benpicco replied:
Hm `netconn_write_vectors_partly()` would have mapped fine to v1 of this PR, but now I don't have an array of snips anymore.

 The data was already copied into a continuous buffer `buf` for the UDP case, so can't we just use that? It would be unfortunate though if we had to copy that again with `NETCONN_COPY`.

yarrick replied:
Yes, sending the `buf` contents in one go like this is fine.

`NETCONN_COPY` was if we did like in v1 (and not used the buf but only forwarded the pointers)

fjmolinas commented:
@miri64 are you OK with this one? If yes could you give it an ACK? I would rather have someone maintaining this system ack this :)

kaspar030 reviewed (changes requested):
I'm still not convinced that introducing sock_tx_snip_t and sock_rx_snip_t is a good idea.

The main pro for that would be to not have to cast R/O `data` to non-const, for the case we'd send on-flash data. Also, this makes it clear from an API perspective that the callee won't ever modify the passed in data. IIUC there was also some unhappyness with the iolist member names.

Reasons for just re-using iolist_t:

1. layers below use iolist_t
2. other modules might use or produce iolist_t that could be passed to a socket send function.
3. instead of one type (iolist_t), there's now three (iolist_t, sock_tx_snip_t, sock_rx_snip_t)
4. if the member names of iolist_t suck, we can change them globally.

I think those three types will cause conversion, type casting, ..., totally outweighing the constness benefit. And I think that constness benefit can be mitigated with clear documentation (like, in the docks for sock_send(..., iolist_t snip): "@note don't modify snip->iol_data!).

I really want this functionality though. Could we, as a compromise, use iolist_t in this PR and postpone the type shedding to later?

kaspar030 commented:
> 3\. instead of one type (iolist_t), there's now three (iolist_t, sock_tx_snip_t, sock_rx_snip_t)

There's also the point that if having distinct rx/tx snips is actually the preferred choice, we should have that for iolist_t.
But, while I don't think that posix is a good example for API design, they get away with only one iovec_t type. It does reduce complexity.

benpicco commented:
Find, I still maintain that the `iolist_t` API is fugly, but if you prefer it that way so be it.

kaspar030 commented:
> Fine, I still maintain that the `iolist_t` API is fugly

you mean because of the member names, or because of the const?

kaspar030 commented:
From my perspective you can squash this!

benpicco commented:
> you mean because of the member names, or because of the const?

both really, e.g.

benpicco commented on tests/lwip_sock_udp/main.c:
```diff
@@ -1288,6 +1288,38 @@ static void test_sock_udp_send6__socketed(void)
+                                          .port = _TEST_PORT_REMOTE };
+
+    const iolist_t tail = {
+        .iol_base = "EFGH",
```
this would now be needlessly placed in RAM if `tail` were not `const` itself

kaspar030 replied:
quick testing shows they are with our cortex-m gcc

kaspar030 reviewed (approved):
ACK.

Code looks fine. Did some basic UDP testing, all good.

benpicco commented:
Thank you for the review!
